cql3/statements/select_statement.cc

```diff
@@ -404,32 +405,58 @@ select_statement::do_execute(query_processor& qp,
 
     auto key_ranges = _restrictions->get_partition_key_ranges(options);
 
+    bool send_tablet = false;
+
+    auto token = dht::token();
+    unsigned shard;
+
+    if (key_ranges.size() > 0 && query::is_single_partition(key_ranges.front())) {
+        token = key_ranges[0].start()->value().as_decorated_key().token();
+        shard = _schema->table().shard_of(token);
+    }
+
+    auto&& table = qp.proxy().local_db().find_column_family(_schema);
+    auto erm = table.get_effective_replication_map();
+    auto tablet_replicas = erm->get_replicas_for_reading(token);
+    auto token_range = erm->get_token_range(token);
+
+    if ((!tablet_replicas.empty() && erm->get_token_metadata_ptr()->get_endpoint_for_host_id(tablet_replicas.front().host) != utils::fb_utilities::get_broadcast_address())
+        || this_shard_id() != shard) {
+        send_tablet = true;
+    }
+
     if (db::is_serial_consistency(options.get_consistency())) {
         if (key_ranges.size() != 1 || !query::is_single_partition(key_ranges.front())) {
              throw exceptions::invalid_request_exception(
                      "SERIAL/LOCAL_SERIAL consistency may only be requested for one partition at a time");
         }
-        unsigned shard = _schema->table().shard_of(key_ranges[0].start()->value().as_decorated_key().token());
         if (this_shard_id() != shard) {
             return make_ready_future<shared_ptr<cql_transport::messages::result_message>>(
-                    qp.bounce_to_shard(shard, std::move(const_cast<cql3::query_options&>(options).take_cached_pk_function_calls()))
-                );
+                qp.bounce_to_shard(shard, std::move(const_cast<cql3::query_options&>(options).take_cached_pk_function_calls()), tablet_replicas, token_range)
```

Comment:
bounce_to_shard() result is handled by the cql server, it's not sent to the driver:

```
        auto* bounce_msg = std::get_if<shared_ptr<messages::result_message::bounce_to_shard>>(&msg);
        if (bounce_msg) {
            return process_on_shard(*bounce_msg, stream, is, client_state, std::move(permit), trace_state, process_fn);
        }
        auto ptr = std::get<cql_server::result_with_foreign_response_ptr>(std::move(msg));
        return make_ready_future<cql_server::result_with_foreign_response_ptr>(std::move(ptr));
```

Reply:
Fixed